knodle/trainer/cosine/cosine.py

```diff
@@ -0,0 +1,496 @@
+        with torch.no_grad():
+            for idx, eval_batch in enumerate(eval_set_loader):
+                input_ids = eval_batch['input_ids'].to(device)
+                attention_mask = eval_batch['attention_mask'].to(device)
+                targets = eval_batch['labels'].to(device)
+
+                y_logits = model(input_ids, attention_mask)['logits']
+                loss_sum += loss_fn(y_logits, targets)
+                y_preds = torch.max(y_logits.cpu(), 1)[1]
+                all_preds.extend(y_preds.numpy())
+                all_y.extend(list(targets.cpu()))
+
+                if fast_mode and idx > num_batches:
+                    break
+
+            classification_score_dict = classification_report(all_y, np.array(all_preds).flatten(), output_dict=True)
+            classification_score_str = classification_report(all_y, np.array(all_preds).flatten(), output_dict=False)
+
+            if verbose:
+                logger_prefix = kwargs['tag'] if 'tag' in kwargs else ''
+                global_step = kwargs['global_step'] if 'global_step' in kwargs else -1
+                self.logger.info(f'{logger_prefix} score at step {global_step}')
+                self.logger.info(classification_score_str)
+
+        return {'score_dict': classification_score_dict,
+                'score_str': classification_score_str,
+                'loss': loss_sum/(len(all_y))}
+
+    def get_optimizer_grouped_parameters(self, config, model):
+        no_decay = ['bias', 'gamma', 'beta', 'LayerNorm.weight']
+        optimizer_parameters = [
+            {'params': [p for n, p in model.named_parameters() if not any(nd in n for nd in no_decay)],
+             'weight_decay': config.weight_decay},
+            {'params': [p for n, p in model.named_parameters() if any(nd in n for nd in no_decay)],
+             'weight_decay': 0.0}
+        ]
+
+        return optimizer_parameters
+
+
+    def calc_loss(self, input, target, loss, device, thresh=0.95, soft=True, conf='max', confreg=0.1):
+        softmax = nn.Softmax(dim=1)
+        target = softmax(target.view(-1, target.shape[-1])).view(target.shape)
+
+        if conf == 'max':
+            weight = torch.max(target, dim=1).values
+            # w = torch.FloatTensor([1 if x == True else 0 for x in weight > thresh]).to(device)
+            # w = torch.tensor(weight > thresh, dtype=torch.float)
+            w = (weight > thresh).float()
+        elif conf == 'entropy':
+            weight = torch.sum(-torch.log(target + 1e-6) * target, dim=1)  # Entropy
+            weight = 1 - weight / np.log(weight.size(-1))
+            # w = torch.FloatTensor([1 if x == True else 0 for x in weight > thresh]).to(device)
+            # w = torch.tensor(weight > thresh, dtype=torch.float)
+            w = (weight > thresh).float()
+        else:
+            raise ValueError(f'conf={conf} is unsupported')
+        target = self.soft_frequency(target, probs=True, soft=soft)
+
+        loss_batch = loss(input, target)
```

Comment:
You compute it on the full batch. Shouldn't this be computed on the subset "C" (samples over treshold, formula 10 in paper)?

Reply:
Hi,
yes, the variable `loss_batch` is indeed computed over the entire batch. However, the final loss `lc` is computed on the subset "C" (line 218 in the current code version).

Reply:
Hi, thanks for clarification.  I think I had problems understanding w / weight. This and line 218 makes sense to me now.